crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -379,14 +654,82 @@ fn load_object_arg<S: Storage>(
     state_view: &S,
     object_owner_map: &mut BTreeMap<ObjectID, Owner>,
     obj_arg: ObjectArg,
-) -> Result<ObjectValue, ExecutionError> {
+) -> Result<InputValue, ExecutionError> {
     match obj_arg {
         ObjectArg::ImmOrOwnedObject((id, _, _))
         | ObjectArg::SharedObject {
             id, mutable: true, ..
-        } => load_object(state_view, object_owner_map, None, id),
+        } => load_object(
+            state_view,
+            object_owner_map,
+            /* imm override */ false,
+            id,
+        ),
         ObjectArg::SharedObject {
             id, mutable: false, ..
-        } => load_object(state_view, object_owner_map, Some(Owner::Immutable), id),
+        } => load_object(
+            state_view,
+            object_owner_map,
+            /* imm override */ true,
+            id,
+        ),
+    }
+}
+
+fn add_additional_write(
+    additional_writes: &mut BTreeMap<ObjectID, AdditionalWrite>,
+    owner: Owner,
+    object_value: ObjectValue,
+) {
+    let ObjectValue {
+        type_,
+        has_public_transfer,
+        contents,
+        ..
+    } = object_value;
+    let bytes = match contents {
+        ObjectContents::Coin(coin) => coin.to_bcs_bytes(),
+        ObjectContents::Raw(bytes) => bytes,
+    };
+    let object_id = MoveObject::id_opt(&bytes).unwrap();
+    additional_writes.insert(object_id, (owner, type_, has_public_transfer, bytes));
+}
+
+fn create_written_object(
+    protocol_config: &ProtocolConfig,
+    input_object_metadata: &BTreeMap<ObjectID, InputObjectMetadata>,
+    loaded_child_objects: &BTreeMap<ObjectID, SequenceNumber>,
+    id: ObjectID,
+    tag: StructTag,
+    has_public_transfer: bool,
+    contents: Vec<u8>,
+    write_kind: WriteKind,
+) -> Result<MoveObject, ExecutionError> {
+    debug_assert_eq!(
+        id,
+        MoveObject::id_opt(&contents).expect("object contents should start with an id")
+    );
+    let metadata_opt = input_object_metadata.get(&id);
+    let loaded_child_version_opt = loaded_child_objects.get(&id);
+    assert_invariant!(
+        metadata_opt.is_none() || loaded_child_version_opt.is_none(),
+        format!("Loaded {id} as a child, but that object was an input object")
+    );
+
+    let old_obj_ver = metadata_opt
+        .map(|metadata| metadata.version)
+        .or_else(|| loaded_child_version_opt.copied());
+
+    debug_assert!((write_kind == WriteKind::Mutate) == old_obj_ver.is_some());
+
+    // safe because `has_public_transfer` was properly determined from the abilities
```

Comment:
Shouldn't this function also be `unsafe` because `has_public_transfer` is just passed in?  This safety comment is not locally true.